File: ruby sketch/lib/frame.rb
class Frame
  attr_reader :pins, :is_over, :strike, :spare, :score, :roll_array

  def initialize
    @pins = 10
    @rolls = 0
    @is_over = false
    @strike = false
    @spare = false
    @score = 0
    @roll_array = []
  end

  def first_roll
    @roll_array[0]
  end

  def roll(hit)
    @pins -= hit
    @rolls += 1
    @roll_array << hit unless @is_over
    check_if_frame_over
  end

  def check_if_frame_over
    if @rolls == 2
      @is_over = true
      @spare = true if @pins == 0
    end
    if @pins == 0 && @rolls == 1
      @is_over = true
      @strike = true
    end
    @score = 10 - @pins
  end
end

File: bowling-challenge-1st-attempt/src/frame.js
function Frame() {
  this._rolls = [];
  this._score = 0;
  this._bonus = 'none';
  this.isOver = false;
};


Frame.prototype.score = function() {
  return this._score;
};

Frame.prototype.addToScore = function(value) {
  this._score += value;
};

Frame.prototype.rolls = function() {
  return this._rolls;
};

Frame.prototype.bonus = function() {
  return this._bonus;
};

Frame.prototype.addRoll = function(roll, frameIndex) {
  if(roll > 10) {
    throw new Error('Input Error: Cannot knock down more than 10 Pins in a roll!')
  } else if(this._score + roll > 10 && frameIndex !== 10) {
    throw new Error('Input Error: Cannot knock down more than 10 Pins per frame!')
  } else {
    this._rolls.push(roll);
    this._score += roll;
  }
};

Frame.prototype.calcBonus = function() {
  if(this._score === 10) {
    switch(this._rolls.length) {
      case 1:
        this._bonus = 'strike'
        break;
      case 2:
        this._bonus = 'spare'
      default:
    };
  }
};
